src/WindowsUtils/DebuggerTest.cpp

```diff
@@ -0,0 +1,97 @@
+
+namespace {
+
+std::filesystem::path GetTestExecutablePath() {
+  static auto path = orbit_base::GetExecutableDir() / "FakeCliProgram.exe";
+  return path;
+}
+
+using orbit_windows_utils::DebugEventListener;
+using orbit_windows_utils::Debugger;
+using orbit_windows_utils::ProcessInfo;
+
+}  // namespace
+
+struct MockListener : public DebugEventListener {
+  MOCK_METHOD(void, OnCreateProcessDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnExitProcessDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnCreateThreadDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnExitThreadDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnLoadDllDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnUnLoadDllDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnBreakpointDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnOutputStringDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnExceptionDebugEvent, (const DEBUG_EVENT& event), (override));
+  MOCK_METHOD(void, OnRipEvent, (const DEBUG_EVENT& event), (override));
+};
+
+TEST(Debugger, LaunchProcess) {
+  constexpr size_t kNumListeners = 10;
+  std::vector<MockListener> mock_listeners(kNumListeners);
+  std::vector<DebugEventListener*> listeners;
+  for (MockListener& mock_listener : mock_listeners) {
+    listeners.push_back(&mock_listener);
+  }
+
+  Debugger debugger(listeners);
+
+  for (MockListener& mock_listener : mock_listeners) {
+    EXPECT_CALL(mock_listener, OnCreateProcessDebugEvent).Times(1);
+    EXPECT_CALL(mock_listener, OnExitProcessDebugEvent).Times(1);
+    EXPECT_CALL(mock_listener, OnBreakpointDebugEvent).Times(1);
+    EXPECT_CALL(mock_listener, OnCreateThreadDebugEvent).Times(testing::AtLeast(1));
+    EXPECT_CALL(mock_listener, OnExitThreadDebugEvent).Times(testing::AtLeast(1));
+    EXPECT_CALL(mock_listener, OnLoadDllDebugEvent).Times(testing::AnyNumber());
+    EXPECT_CALL(mock_listener, OnUnLoadDllDebugEvent).Times(testing::AnyNumber());
+    EXPECT_CALL(mock_listener, OnOutputStringDebugEvent).Times(testing::AnyNumber());
+    EXPECT_CALL(mock_listener, OnExceptionDebugEvent).Times(testing::AnyNumber());
+    EXPECT_CALL(mock_listener, OnRipEvent).Times(testing::AnyNumber());
+  }
+
+  const std::string kArguments = "--sleep_for_ms=20";
+  auto result = debugger.Start(GetTestExecutablePath(), /*working_directory=*/"", kArguments);
+  ASSERT_TRUE(result.has_value());
+
+  Debugger::StartInfo& start_info = result.value();
+  std::string expected_command_line = GetTestExecutablePath().string();
+  if (!kArguments.empty()) expected_command_line += absl::StrFormat(" %s", kArguments);
+  EXPECT_STREQ(start_info.command_line.c_str(), expected_command_line.c_str());
+
+  // Wait for debugger to automatically detach upon process termination.
+  debugger.Wait();
```

Comment:
I think we should (if possible) also test that the process is terminated here. And probably also that the process is running before the call.

Reply:
Added test that the process is not running after "Wait". Testing that the process is running before the call doesn't work in this particular test because the process could exit quickly before we have time to test that it is running. It is however tested in the new "Detach" test which spawns a process that sleeps forever.